include/ginkgo/core/distributed/assembly_helpers.hpp

```diff
@@ -0,0 +1,56 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_DISTRIBUTED_ASSEMBLY_HELPERS_HPP_
+#define GKO_PUBLIC_CORE_DISTRIBUTED_ASSEMBLY_HELPERS_HPP_
+
+
+#include <ginkgo/config.hpp>
+
+
+#if GINKGO_BUILD_MPI
+
+
+#include <ginkgo/core/base/device_matrix_data.hpp>
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/base/range.hpp>
+
+
+namespace gko {
+namespace experimental {
+namespace distributed {
+
+template <typename LocalIndexType, typename GlobalIndexType>
+class Partition;
+
+
+/**
+ * Assembles device_matrix_data entries owned by this MPI rank from other ranks
+ * and communicates entries located on this MPI rank owned by other ranks to
+ * their respective owners. This can be useful e.g. in a finite element code
+ * where each rank assembles a local contribution to a global system matrix and
+ * the global matrix has to be assembled by summing up the local contributions
+ * on rank boundaries.
+ *
+ * @param comm the communicator used to assemble the global matrix.
+ * @param input the device_matrix_data structure.
+ * @param partition the partition used to determine owndership.
+ *
+ * @return the globally assembled device_matrix_data structure for this MPI
+ * rank.
+ */
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+device_matrix_data<ValueType, GlobalIndexType> assemble(
+    mpi::communicator comm,
+    const device_matrix_data<ValueType, GlobalIndexType>& input,
+    ptr_param<const Partition<LocalIndexType, GlobalIndexType>> partition);
```

Comment:
Maybe rename it to 
```suggestion
    ptr_param<const Partition<LocalIndexType, GlobalIndexType>> row_partition);
```
since only the ownership of the rows is relevant here. Alternatively, mention it in the doc.

Reply:
If this is meant to be used in conjunction with a row-partitioned matrix, we should probably include that in the name. There are three types of distributions we can have (symmetric partitioning assumed): row-partitioned, column-partitioned and overlapping. This is probably for row-partitioning, we would need a slightly different routine for column-partitioned operations, and overlapping works without explicit assembly.

Reply:
Just for my understanding, a row- and column partitioned matrix would mean that the matrix is decomposed into blocks, and each block is owned by a different process, right? In that case, overlapping might not be the right word to describe it.